Improve behavior when a CHPL_LLVM=none build is used when CHPL_LLVM!=none

bradcray opened the pull request:
When using a compiler built with `CHPL_LLVM=none` in an environment
that has `CHPL_LLVM` set or inferred to `system` or `bundled`, the
traditional error message has been short, but not particularly clear.
I keep hitting this error and find the current message difficult
to understand, particularly putting myself in the shoes of someone
who, say, built Chapel with the quickstart build and then is trying to
use it from a new shell that may not set CHPL_LLVM=none.  Or
someone piggy-backing off of someone else's installation that was
built without LLVM.

This PR improves things by:

* setting `CHPL_LLVM=none` in the compiler's environment in the event
  that it is unset and the compiler was built without LLVM support; this
  happens early in compilation so that when printchplenv is invoked, an
  appropriate back-end compiler will be selected.  I also added a note
  to `--print-chpl-settings` to indicate that this setting had been made
  by the compiler (and why).
* improving the existing error message so that if someone has explicitly
  requested the llvm back-end through `--target-compiler`, `CHPL_TARGET_COMPILER`,
  or `CHPL_LLVM` with a compiler that was built without LLVM, they get
  a better message than they used to.
* reordering how we do things early in the compiler so that things like `--help`
  are handled before we print errors like this one.  That way, someone running
  `chpl --help` won't immediately get an error about how they can't compile
  when they weren't actually trying to.

Resolves #18073.
Resolves #18197.

mppf commented:
I wonder if the error message should talk about `CHPL_TARGET_COMPILER=llvm` vs `CHPL_LLVM`.

mppf commented:
(because we can select the C backend by saying something like `CHPL_TARGET_COMPILER=gnu` even if CHPL_LLVM is not none, e.g.; or we could set `CHPL_TARGET_COMPILER=llvm` even if CHPL_LLVM were none and expect a reasonable error).

bradcray commented:
Oh, that's an attractive idea.

bradcray commented:
@mppf: So maybe:

```
error: You have requested 'llvm' as the target compiler, but this copy of
       'chpl' was built without LLVM support.  Either set 'CHPL_LLVM=none'
       or re-build your compiler with LLVM enabled.
```

?

mppf commented:
> error: You have requested 'llvm' as the target compiler, but this copy of 'chpl' was built without LLVM support.  Either set 'CHPL_LLVM=none' or re-build your compiler with LLVM enabled.

The part starting with "Either set" still talks about CHPL_LLVM though.  Perhaps it could just say

> error: You have requested 'llvm' as the target compiler, but this copy of 'chpl' was built without LLVM support.  Consider rebuilding your compiler with LLVM enabled.

Otherwise I fear we will muddy the error message with the difference between CHPL_LLVM and CHPL_TARGET_COMPILER.

bradcray commented:
I don't like having the error imply that "rebuild" is the only/main option, because I think the typical user could be using an installation of `chpl` that someone else built and may have no idea how to build a copy of Chapel.  I went for the CHPL_LLVM=none approach because (a) I don't think we're planning on retiring it (is that right?), (b) it avoids having to make a best guess as to what C compiler they might want to select as their target compiler instead, and (c) it's what I use.  What about it makes you wince?

bradcray commented:
(one reservation I have about the revised error messages is "you have requested 'llvm' as your target compiler" could be vague to new / naive users as well.  I could imagine a new user running a compiler somewhere and saying "What?  No, I didn't request anything, you just inferred something.")

mppf commented:
> I don't like having the error imply that "rebuild" is the only/main option, because I think the typical user could be using an installation of chpl that someone else built and may have no idea how to build a copy of Chapel. I went for the CHPL_LLVM=none approach because (a) I don't think we're planning on retiring it (is that right?), (b) it avoids having to make a best guess as to what C compiler they might want to select as their target compiler instead, and (c) it's what I use. What about it makes you wince?

Just that setting `CHPL_LLVM=none` is not enough if somebody has set `CHPL_TARGET_COMPILER=llvm` - they would have to change/unset `CHPL_TARGET_COMPILER` too. But maybe that is a weird enough thing to do that we are better off ignoring it (?).

bradcray commented:
> Just that setting CHPL_LLVM=none is not enough if somebody has set CHPL_TARGET_COMPILER=llvm - they would have to change/unset CHPL_TARGET_COMPILER too. But maybe that is a weird enough thing to do that we are better off ignoring it (?).

Oh, I see.  Or maybe we can rely on someone who's explicitly set that to know what they're doing and unset it?  Or we could add another option like "request a different target compiler".

Thinking about this over lunch, I'm thinking maybe our approach here should be different.  The case when I hit this most is:
* I'm using a compiler that wasn't built with LLVM enabled
* I haven't said anything about LLVM or target compilers
where today, our scripting says "Oh, you're on a system with a system installation of LLVM (or a built bundled version), so I'm going to infer `CHPL_LLVM=system` (or bundled) on your behalf and tell the compiler to use that."  But it seems silly to have our compiler infer that it should be using LLVM and then complain that it can't.  Put another way, I'd expect a compiler with LLVM not enabled to always infer `CHPL_LLVM=none` / `CHPL_TARGET_COMPILER=[appropriate C compiler]` if they aren't set.

But I imagine that implementing this is a bit sticky because it essentially requires either (a) our chplenv scripts to ask stuff from the compiler (which I don't think we have a precedent for?) or (b) having the compiler ignore stuff from the chplenv scripts if it's inferred rather than set (if the compiler can even tell), which would have the effect of making `chpl --print-chpl-settings` differ from `util/printchplenv`.

Thoughts on that?

mppf commented:
> having the compiler ignore stuff from the chplenv scripts if it's inferred rather than set (if the compiler can even tell), which would have the effect of making chpl --print-chpl-settings differ from util/printchplenv.

Of the two, this is the one that makes sense to me. But I agree with you that it'd be new and different.

Perhaps the right solution is to just make the error we are talking about into a warning - and if the compiler is built without LLVM support it just prints the warning and sets `CHPL_LLVM=none` for itself. That behavior seems straightforward and not confusing to me.

bradcray commented:
> Of the two, this is the one that makes sense to me. But I agree with you that it'd be new and different.
> 
> Perhaps the right solution is to just make the error we are talking about into a warning - and if the compiler is built without LLVM support it just prints the warning and sets CHPL_LLVM=none for itself. That behavior seems straightforward and not confusing to me.

Thinking about this a few minutes more, I don't think it would work, or at least not easily.  Specifically, in addition to setting CHPL_LLVM=none, it would have to infer a reasonable back-end compiler because the scripting will tell it `llvm` as the back-end compiler.  And that feels more difficult to drag from the scripts into the compiler.  Unless we were to set `CHPL_LLVM=none` in the environment if (a) if it is unset and (b) before we even call the scripts in the event that the compiler was built without LLVM support (?).

I'm starting to think I should fork this off into its own issue in any case.

bradcray commented:
@mppf: This has now changed pretty significantly since your original review (and I've updated the OP to indicate that, and how).  I know you're probably trying to wrap stuff up before vacation, but do you have time to re-review, or should I ask someone else?  Thanks.

mppf commented on compiler/main/driver.cpp:
```diff
@@ -506,6 +506,21 @@ static void setupChplHome(const char* argv0) {
+  // set CHPL_LLVM to 'none' if it isn't already set
+  if (setenv("CHPL_LLVM", "none", 0) != 0) {
+    INT_FATAL("Problem setting CHPL_LLVM");
+  }
```
I was expecting this to call `getenv` first to see if CHPL_LLVM is set, and if not, do the `setenv` *and print a warning*.

mppf reviewed (approved):
I'm happy with the wording improvement for the error but am hoping for a warning when we infer CHPL_LLVM=none in the compiler based on how it was built (since this is not generally the default).

bradcray commented:
I saw your suggestion to include a warning, but as someone who hits this use case frequently, I think I'd find the warning to just be an annoyance rather than helpful / reassuring in any way.  For example, it would read to me a bit like "I know you didn't do anything to request an llvm compile, and you can't actually do an llvm compile, but I'm going to warn you that I'm not planning on doing an llvm compile even though your environment suggests I could if you / someone else were to rebuild the compiler in a different way."  So to help me understand your perspective, what is the scenario in which a warning would be helpful rather than TMI?

mppf commented:
The reason I imagine the warning being useful is just that it addresses the mismatch between `printchplenv` and `chpl --print-chpl-settings`. (I.e., user runs `printchplenv` and sees they are using the LLVM backend, does performance testing or something, but didn't realize they weren't using the C backend all along). If we leave out the warning, we might want to update the new issue template to ask for `chpl --print-chpl-settings` rather than `$CHPL_HOME/util/printchplenv --anonymize`. (Not sure if we need `chpl --print-chpl-settings-anonymize` or something).

bradcray commented:
> If we leave out the warning, we might want to update the new issue template to ask for chpl --print-chpl-settings rather than $CHPL_HOME/util/printchplenv --anonymize. (Not sure if we need chpl --print-chpl-settings-anonymize or something).

Independent of this, I'd been thinking (now that I remember it exists again) that we should ask users to report their environment using `chpl --print-chpl-settings` anyway, as it seems easier for them (just another invocation of the compiler; they don't have to figure out where `printchplenv` lives on the XC or homebrew install, say); and maybe there are other subtle differences prior to this one where the compiler is setting or inferring something (as a simple example: maybe due to command-line arguments they forgot they're throwing in a Makefile example) that wouldn't be captured in a `printchplenv` run.  So for example, rather than asking them to literally run `chpl --print-chpl-settings`, we could say "re-run your compilation step with `--print-chpl-settings` tacked onto the end, and then paste the output in your issue or email."  I hadn't thought about the `anonymize` aspect of it, though...  Maybe we could just have `--print-chpl-settings` always do `--anonymize` for users, though, and non-anonymize for developers?

Back on this issue, though:  We could potentially have `--print-chpl-settings` print something special in the case where the compiler has set CHPL_LLVM to none which would address your concern while still keeping the typical compilation quiet.  I was already feeling a little weird that it puts the `*` next to CHPL_LLVM which might cause a savvy user to say "but that's _not_ set in my environment?!".  So it could be something like:

```
...normal stuff...
CHPL_LLVM: none *
...normal stuff...
---
* note that CHPL_LLVM was automatically set to `none` by the 'chpl' compiler since it was built without LLVM support enabled.
```

mppf commented:
Yeah, these steps are OK with me.

bradcray commented:
I created #18206 to capture the notion of moving away from `printchplenv` when we ask users for things.

bradcray commented:
https://github.com/chapel-lang/chapel/pull/18194/commits/5b5eeea0a1f4f6c305b2cf7031bde5cf526c14c8 adds the note to `--print-chpl-settings`

ronawho commented on compiler/main/driver.cpp:
```diff
@@ -1595,9 +1626,17 @@ static void postprocess_args() {
+// errors rather than doing what the user said (for which these
+// checks don't apply because we're not going to compile anything).
+//
+static void checkStuff() {
```
`checkStuff()` doesn't seem like a very informative name to me, but I suppose it's no worse than `printStuff()`

bradcray replied:
That is where I took my inspiration from.  :D

bradcray replied:
I renamed this to `validateSettings()` in the most recent commit.

ronawho commented on compiler/main/driver.cpp:
```diff
@@ -506,6 +508,24 @@ static void setupChplHome(const char* argv0) {
+    if (setenv("CHPL_LLVM", "none", 0) != 0) {
+      INT_FATAL("Problem setting CHPL_LLVM");
+    }
+    setChplLLVM = true;
```
Is there any conflict here if the user throws `--llvm` (or maybe that's been replaced recently, so whatever has replaced it.) I think it's a user error if they're throwing `--llvm` for a compiler built without it, but I'm not sure if it changes the error message in an unhelpful way.

ronawho replied:
Ah, based on the PR description and the code you updated in `checkLLVMCodeGen`, this code must be called after that `checkLLVMCodeGen`, so I think nevermind this question.

bradcray replied:
`--llvm` is deprecated but still supported for this release, so throwing it results in a deprecation error and my new one:

```
warning: --llvm is deprecated -- please use --target-compiler=llvm
error: You have requested 'llvm' as the target compiler, but this copy of
       'chpl' was built without LLVM support.  Either select a different
       target compiler or re-build your compiler with LLVM enabled.
```

Similarly, using `--target-compiler=llvm` (or the equivalent CHPL_TARGET_COMPILER env var) results in:

```
error: You have requested 'llvm' as the target compiler, but this copy of
       'chpl' was built without LLVM support.  Either select a different
       target compiler or re-build your compiler with LLVM enabled.
```

Does that match what you were imagining / hoping for?

ronawho replied:
Yup, thanks.

ronawho commented on compiler/main/driver.cpp:
```diff
@@ -506,6 +508,24 @@ static void setupChplHome(const char* argv0) {
   }
 }
 
+#ifndef HAVE_LLVM
```
Might be a little cleaner to stuff this `ifndef` into the the function and avoid the `ifndef` at the call site? No real preference here, so I'm fine leaving as-is.

bradcray replied:
Did this.

ronawho commented on compiler/main/driver.cpp:
```diff
@@ -295,6 +295,8 @@ bool fCompilerLibraryParser = false;
 
 chpl::Context* gContext = nullptr;
 
+static bool setChplLLVM = false;
```
Minor preference to call this `setChplLLVMEnv` or `compilerOverrodeChplLLVM` or something to indicate that this is not reflecting a user setting, but something the compiler did. (When I read this initially I was assuming it meant the user threw `--llvm`)

ronawho reviewed (approved):
Had some minor nitpicks, but I'd be happy merging as-is if those don't seem worth addressing.
